rasa/engine/caching.py

```diff
@@ -0,0 +1,373 @@
+        incompatible_entries = self._find_incompatible_cache_entries()
+
+        for entry in incompatible_entries:
+            self._delete_cached_result(entry)
+
+        self._delete_incompatible_entries_from_cache(incompatible_entries)
+
+        logger.debug(
+            f"Deleted {len(incompatible_entries)} from disk as their version "
+            f"is smaller than the minimum compatible version "
+            f"('{MINIMUM_COMPATIBLE_VERSION}')."
+        )
+
+    def _find_incompatible_cache_entries(self) -> List[TrainingCache.CacheEntry]:
+        with self._sessionmaker() as session:
+            query_for_cache_entries = sa.select(self.CacheEntry)
+            all_entries: List[TrainingCache.CacheEntry] = session.execute(
+                query_for_cache_entries
+            ).scalars().all()
+
+        return [
+            entry
+            for entry in all_entries
+            if version.parse(MINIMUM_COMPATIBLE_VERSION)
+            > version.parse(entry.rasa_version)
+        ]
+
+    def _delete_incompatible_entries_from_cache(
+        self, incompatible_entries: List[TrainingCache.CacheEntry]
+    ) -> None:
+        incompatible_fingerprints = [
+            entry.fingerprint_key for entry in incompatible_entries
+        ]
+        with self._sessionmaker.begin() as session:
+            delete_query = sa.delete(self.CacheEntry).where(
+                self.CacheEntry.fingerprint_key.in_(incompatible_fingerprints)
+            )
+            session.execute(delete_query)
+
+    @staticmethod
+    def _delete_cached_result(entry: TrainingCache.CacheEntry) -> None:
+        if entry.result_location and Path(entry.result_location).is_dir():
+            shutil.rmtree(entry.result_location)
+
+    def cache_output(
+        self,
+        fingerprint_key: Text,
+        output: Any,
+        output_fingerprint: Text,
+        model_storage: ModelStorage,
+    ) -> None:
+        """Adds the output to the cache.
+
+        If the output is of type `Cacheable` the output is persisted to disk in addition
+        to its fingerprint.
+
+        Args:
+            fingerprint_key: The fingerprint key serves as key for the cache. Graph
+                components can use their fingerprint key to lookup fingerprints of
+                previous training runs.
+            output: The output. The output is only cached to disk if it's of type
+                `Cacheable`.
+            output_fingerprint: The fingerprint of they output. This can be used
+                to lookup potentially persisted outputs on disk.
+            model_storage: Required for caching `Cacheable` instances. E.g. `Resource`s
+                use that to copy data from the model storage to the cache.
+        """
+        if self._is_disabled():
+            return
+
+        cache_dir, output_type = None, None
+        if isinstance(output, Cacheable):
+            cache_dir, output_type = self._cache_output_to_disk(output, model_storage)
```

Comment:
If the next step fails should we clean this up? Otherwise it will never be removed right?

Reply:
I'll turn around the order that should boil down to the same effect 👍🏻

Reply:
handled 👍🏻